mlir/lib/Dialect/Presburger/Set.cpp

```diff
@@ -0,0 +1,396 @@
+  auto addInequalityFromEquality = [&](const ArrayRef<int64_t> &eq,
+                                       bool negated, bool strict) {
+    SmallVector<int64_t, 64> coeffs =
+        inequalityFromEquality(eq, negated, strict);
+    B.addInequality(coeffs);
+    simplex.addInequality(coeffs);
+  };
+  auto recurseWithInequalityFromEquality = [&, i](const ArrayRef<int64_t> &eq,
+                                                  bool negated, bool strict) {
+    size_t snap = simplex.getSnapshot();
+    addInequalityFromEquality(eq, negated, strict);
+
+    subtractRecursively(B, simplex, S, i + 1, result);
+
+    // TODO check if this removes the right inequality
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+  };
+
+  size_t addedIneqs = 0;
+
+  for (size_t j = 0; j < S_i.getNumEqualities(); j++) {
+    // The first inequality is positive and the second is negative, of which
+    // we need the complements (strict negative and strict positive).
+    // TODO reimplement the heuristics
+    const auto &eq = S_i.getEquality(j);
+    recurseWithInequalityFromEquality(eq, true, true);
+    recurseWithInequalityFromEquality(eq, false, true);
+
+    addInequalityFromEquality(eq, false, false);
+    addInequalityFromEquality(eq, true, false);
+    addedIneqs += 2;
+  }
+
+  // offset = 2 * S_i.getNumEqualities();
+  for (size_t j = 0; j < S_i.getNumInequalities(); j++) {
+    /*if (isMarkedRedundant[offset + j])
+      continue;*/
+    const auto &ineq = S_i.getInequality(j);
+
+    size_t snap = simplex.getSnapshot();
+
+    SmallVector<int64_t, 64> complement = complementIneq(ineq);
+
+    B.addInequality(complement);
+    simplex.addInequality(complement);
+    subtractRecursively(B, simplex, S, i + 1, result);
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+
+    B.addInequality(ineq);
+    addedIneqs++;
+    simplex.addInequality(ineq);
+  }
+
+  for (size_t i = 0; i < addedIneqs; i++)
+    B.removeInequality(B.getNumInequalities() - 1);
+
+  // TODO benchmark technically we can probably drop this as the caller will
+  // rollback. See if it makes much of a difference. Only the last rollback
+  // would be eliminated by this.
+  simplex.rollback(initialSnap);
+}
+
+// Returns the set difference B - S.
+PresburgerSet PresburgerSet::subtract(FlatAffineConstraints c,
+                                      const PresburgerSet &set) {
+  assert(set.getNumDims() == c.getNumDimIds() &&
+         set.getNumSyms() == c.getNumSymbolIds() &&
+         "Sets to be subtracted have different dimensionality");
+  if (c.isEmptyByGCDTest())
+    return PresburgerSet::makeEmptySet(c.getNumDimIds(), c.getNumSymbolIds());
+
+  if (set.isUniverse())
+    return PresburgerSet::makeEmptySet(set.getNumDims(), set.getNumSyms());
+  if (set.isMarkedEmpty())
+    return PresburgerSet(set.getNumDims(), set.getNumSyms());
+
+  PresburgerSet result(c.getNumDimIds());
```

Comment:
Need to pass the number of symbols here.

Reply:
Yes. With the other changes it's now possible to reach this with `set.getNumSyms() != 0`